File: packages/shared/src/validation.ts
export const PHONE_INPUT_PLACEHOLDER = "+375339177444";

const PHONE_PATTERN = /^(\+\d{10,12}|\d{11,13})$/;
const PHONE_LETTERS_PATTERN = /[A-Za-zА-Яа-яЁё]/u;
const PHONE_ALLOWED_CHARS_PATTERN = /^[\d+\s\-()]+$/;
const PERSON_NAME_PATTERN = /^[A-Za-zА-Яа-яЁё]+(?:[ \-][A-Za-zА-Яа-яЁё]+)*$/u;
const CAR_MODEL_PATTERN = /^[A-Za-zА-Яа-яЁё0-9.\- ]+$/u;

export type ValidationResult =
  | { ok: true; value: string }
  | { ok: false; error: string };

export function normalizePhone(value: string): string {
  const trimmed = value.trim();
  const hasPlus = trimmed.startsWith("+");
  const digits = trimmed.replace(/\D/g, "");
  return hasPlus ? `+${digits}` : digits;
}

export function normalizePersonName(value: string): string {
  return value.trim().replace(/\s+/g, " ");
}

export function normalizeCarModel(value: string): string {
  return value.trim().replace(/\s+/g, " ");
}

export function validatePhone(raw: string): ValidationResult {
  const trimmed = raw.trim();

  if (!trimmed) {
    return { ok: false, error: "Укажите телефон" };
  }

  if (PHONE_LETTERS_PATTERN.test(trimmed)) {
    return { ok: false, error: "В номере не должно быть букв" };
  }

  if (!PHONE_ALLOWED_CHARS_PATTERN.test(trimmed)) {
    return { ok: false, error: "Неверный формат телефона" };
  }

  const plusCount = (trimmed.match(/\+/g) ?? []).length;
  if (plusCount > 1 || (plusCount === 1 && !trimmed.startsWith("+"))) {
    return { ok: false, error: "«+» только в начале номера" };
  }

  const normalized = normalizePhone(trimmed);
  const digitCount = normalized.replace(/\D/g, "").length;

  if (digitCount < 11) {
    return { ok: false, error: "Номер слишком короткий" };
  }

  if (digitCount > 13) {
    return { ok: false, error: "Номер слишком длинный" };
  }

  if (!PHONE_PATTERN.test(normalized)) {
    return { ok: false, error: "Неверный формат телефона" };
  }

  return { ok: true, value: normalized };
}

export function validatePersonName(raw: string): ValidationResult {
  const normalized = normalizePersonName(raw);

  if (!normalized) {
    return { ok: false, error: "Укажите имя" };
  }

  if (normalized.length < 2) {
    return { ok: false, error: "Минимум 2 символа" };
  }

  if (normalized.length > 100) {
    return { ok: false, error: "Слишком длинное имя" };
  }

  if (/\d/.test(normalized)) {
    return { ok: false, error: "Имя не может содержать цифры" };
  }

  if (!PERSON_NAME_PATTERN.test(normalized)) {
    return { ok: false, error: "Только буквы, пробел и дефис" };
  }

  return { ok: true, value: normalized };
}

export function validateCarModel(raw: string): ValidationResult {
  const normalized = normalizeCarModel(raw);

  if (!normalized) {
    return { ok: false, error: "Укажите модель авто" };
  }

  if (normalized.length < 2) {
    return { ok: false, error: "Минимум 2 символа" };
  }

  if (normalized.length > 100) {
    return { ok: false, error: "Слишком длинная модель" };
  }

  if (!CAR_MODEL_PATTERN.test(normalized)) {
    return { ok: false, error: "Недопустимые символы в модели" };
  }

  if (/\s{2,}/.test(raw.trim())) {
    return { ok: false, error: "Уберите лишние пробелы" };
  }

  if (!/[A-Za-zА-Яа-яЁё]/u.test(normalized)) {
    return { ok: false, error: "Пример: BMW 5 Series" };
  }

  return { ok: true, value: normalized };
}

export function validateOptionalCarModel(raw: string): ValidationResult {
  const normalized = normalizeCarModel(raw);
  if (!normalized) {
    return { ok: true, value: "" };
  }
  return validateCarModel(normalized);
}

File: backend/app/validation.py
import re

PHONE_PATTERN = re.compile(r"^(\+\d{10,12}|\d{11,13})$")
PHONE_LETTERS_PATTERN = re.compile(r"[A-Za-zА-Яа-яЁё]", re.UNICODE)
PHONE_ALLOWED_CHARS_PATTERN = re.compile(r"^[\d+\s\-()]+$")
PERSON_NAME_PATTERN = re.compile(
    r"^[A-Za-zА-Яа-яЁё]+(?:[ \-][A-Za-zА-Яа-яЁё]+)*$",
    re.UNICODE,
)
CAR_MODEL_PATTERN = re.compile(r"^[A-Za-zА-Яа-яЁё0-9.\- ]+$", re.UNICODE)


def normalize_phone(value: str) -> str:
    trimmed = value.strip()
    has_plus = trimmed.startswith("+")
    digits = re.sub(r"\D", "", trimmed)
    return f"+{digits}" if has_plus else digits


def normalize_person_name(value: str) -> str:
    return re.sub(r"\s+", " ", value.strip())


def normalize_car_model(value: str) -> str:
    return re.sub(r"\s+", " ", value.strip())


def validate_phone_number(value: str) -> str:
    trimmed = value.strip()
    if not trimmed:
        raise ValueError("Укажите телефон")
    if PHONE_LETTERS_PATTERN.search(trimmed):
        raise ValueError("В номере не должно быть букв")
    if not PHONE_ALLOWED_CHARS_PATTERN.fullmatch(trimmed):
        raise ValueError("Неверный формат телефона")

    plus_count = trimmed.count("+")
    if plus_count > 1 or (plus_count == 1 and not trimmed.startswith("+")):
        raise ValueError("«+» только в начале номера")

    normalized = normalize_phone(trimmed)
    digit_count = len(re.sub(r"\D", "", normalized))
    if digit_count < 11:
        raise ValueError("Номер слишком короткий")
    if digit_count > 13:
        raise ValueError("Номер слишком длинный")
    if not PHONE_PATTERN.fullmatch(normalized):
        raise ValueError("Неверный формат телефона")
    return normalized


def validate_person_name(value: str) -> str:
    normalized = normalize_person_name(value)
    if not normalized:
        raise ValueError("Укажите имя")
    if len(normalized) < 2:
        raise ValueError("Минимум 2 символа")
    if len(normalized) > 100:
        raise ValueError("Слишком длинное имя")
    if re.search(r"\d", normalized):
        raise ValueError("Имя не может содержать цифры")
    if not PERSON_NAME_PATTERN.fullmatch(normalized):
        raise ValueError("Только буквы, пробел и дефис")
    return normalized


def validate_car_model(value: str, *, required: bool = True) -> str:
    normalized = normalize_car_model(value)
    if not normalized:
        if required:
            raise ValueError("Укажите модель авто")
        return ""
    if len(normalized) < 2:
        raise ValueError("Минимум 2 символа")
    if len(normalized) > 100:
        raise ValueError("Слишком длинная модель")
    if not CAR_MODEL_PATTERN.fullmatch(normalized):
        raise ValueError("Недопустимые символы в модели")
    if re.search(r"\s{2,}", value.strip()):
        raise ValueError("Уберите лишние пробелы")
    if not re.search(r"[A-Za-zА-Яа-яЁё]", normalized):
        raise ValueError("Пример: BMW 5 Series")
    return normalized
